File: recruitpro-api/app/services/interview_round_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from app.repositories.interview_round_repository import InterviewRoundRepository

from app.schemas.interview_round_schema import (
    InterviewRoundCreate,
    InterviewRoundUpdate,
)


class InterviewRoundService:

    def __init__(self):

        self.repository = InterviewRoundRepository()

    # -------------------------
    # Permission Check
    # -------------------------
    def check_permission(
        self,
        current_user: dict,
        permission: str,
    ):

        user_permissions = current_user.get(
            "permissions",
            [],
        )

        if permission not in user_permissions:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You do not have permission to perform this action",
            )

    # -------------------------
    # Get All Interview Rounds
    # -------------------------
    def get_all_interview_rounds(
        self,
        db: Session,
        current_user: dict,
        search: str = "",
        sort_by: str = "RoundName",
        order: str = "asc",
        page: int = 1,
        page_size: int = 10,
    ):

        self.check_permission(
            current_user,
            "VIEW_INTERVIEW_ROUND",
        )

        company_id = current_user.get("company_id")

        if not company_id:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Company information not found for the current user",
            )

        return self.repository.get_all(
            db=db,
            company_id=company_id,
            search=search,
            sort_by=sort_by,
            order=order,
            page=page,
            page_size=page_size,
        )

    # -------------------------
    # Get Interview Round By Id
    # -------------------------
    def get_interview_round_by_id(
        self,
        db: Session,
        interview_round_id: int,
        current_user: dict,
    ):

        self.check_permission(
            current_user,
            "VIEW_INTERVIEW_ROUND",
        )

        interview_round = self.repository.get_by_id(
            db,
            interview_round_id,
        )

        if not interview_round:

            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview Round not found",
            )

        # -------------------------
        # Company Ownership Check
        # -------------------------

        if interview_round.CompanyId != current_user["company_id"]:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not authorized to access this interview round",
            )

        return interview_round

    # -------------------------
    # Create Interview Round
    # -------------------------
    def create_interview_round(
        self,
        db: Session,
        interview_round: InterviewRoundCreate,
        current_user: dict,
    ):

        self.check_permission(
            current_user,
            "CREATE_INTERVIEW_ROUND",
        )

        # Automatically get company
        # from logged-in user

        company_id = current_user.get("company_id")

        if not company_id:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Company information not found for the current user",
            )

        return self.repository.create(
            db=db,
            interview_round=interview_round,
            company_id=company_id,
            current_user=current_user,
        )

    # -------------------------
    # Update Interview Round
    # -------------------------
    def update_interview_round(
        self,
        db: Session,
        interview_round_id: int,
        interview_round: InterviewRoundUpdate,
        current_user: dict,
    ):

        self.check_permission(
            current_user,
            "UPDATE_INTERVIEW_ROUND",
        )

        existing_interview_round = self.repository.get_by_id(
            db,
            interview_round_id,
        )

        if not existing_interview_round:

            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview Round not found",
            )

        # -------------------------
        # Company Ownership Check
        # -------------------------

        if existing_interview_round.CompanyId != current_user["company_id"]:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not authorized to update this interview round",
            )

        # CompanyId is never changed
        # during update

        return self.repository.update(
            db=db,
            interview_round_id=interview_round_id,
            interview_round=interview_round,
            current_user=current_user,
        )

    # -------------------------
    # Delete Interview Round
    # -------------------------
    def delete_interview_round(
        self,
        db: Session,
        interview_round_id: int,
        current_user: dict,
    ):

        self.check_permission(
            current_user,
            "DELETE_INTERVIEW_ROUND",
        )

        interview_round = self.repository.get_by_id(
            db,
            interview_round_id,
        )

        if not interview_round:

            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview Round not found",
            )

        # -------------------------
        # Company Ownership Check
        # -------------------------

        if interview_round.CompanyId != current_user["company_id"]:

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not authorized to delete this interview round",
            )

        self.repository.delete(
            db,
            interview_round_id,
        )

        return {"message": "Interview Round deleted successfully"}

File: recruitpro-ui/src/app/services/interview-round.service.ts
import { Injectable } from '@angular/core';
import {
    HttpClient,
    HttpParams
} from '@angular/common/http';

import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root'
})
export class InterviewRoundService {

    private apiUrl =
        'http://127.0.0.1:8000/interview-rounds';

    constructor(
        private http: HttpClient
    ) { }

    // ==========================
    // Get All Interview Rounds
    // ==========================

    getInterviewRounds(
        search: string = '',
        companyId: number | null = null,
        sortBy: string = 'RoundName',
        order: string = 'asc',
        page: number = 1,
        pageSize: number = 10
    ): Observable<any> {

        let params = new HttpParams()
            .set('search', search)
            .set('sort_by', sortBy)
            .set('order', order)
            .set('page', page.toString())
            .set('page_size', pageSize.toString());

        if (companyId !== null) {
            params = params.set(
                'company_id',
                companyId.toString()
            );
        }

        return this.http.get<any>(
            `${this.apiUrl}/`,
            { params }
        );
    }

    // ==========================
    // Get Interview Round By Id
    // ==========================

    getInterviewRoundById(
        id: number
    ): Observable<any> {

        return this.http.get<any>(
            `${this.apiUrl}/${id}`
        );
    }

    // ==========================
    // Add Interview Round
    // ==========================

    addInterviewRound(
        data: any
    ): Observable<any> {

        return this.http.post<any>(
            `${this.apiUrl}/`,
            data
        );
    }

    // ==========================
    // Update Interview Round
    // ==========================

    updateInterviewRound(
        id: number,
        data: any
    ): Observable<any> {

        return this.http.put<any>(
            `${this.apiUrl}/${id}`,
            data
        );
    }

    // ==========================
    // Delete Interview Round
    // ==========================

    deleteInterviewRound(
        id: number
    ): Observable<any> {

        return this.http.delete<any>(
            `${this.apiUrl}/${id}`
        );
    }
}